src/riak_kv_replrtq_peer.erl

```diff
@@ -0,0 +1,240 @@
+        false ->
+            {ok, #state{}}
+    end.
+
+handle_call({update_discovery, QueueName}, _From, State) ->
+    case lists:keyfind(QueueName, 1, State#state.discovery_peers) of
+        false ->
+            lager:info("Type=~w discovery for unconfigured QueueName=~w",
+                [update, QueueName]),
+            {reply, false, State};
+        {QueueName, PeerInfo} ->
+            R = prompt_discovery(QueueName, PeerInfo, update),
+            {reply, R, State}
+    end;
+handle_call({update_workers, WorkerCount, PerPeerLimit}, _From, State) ->
+    case riak_kv_replrtq_snk:get_worker_counts() of
+        {WorkerCount, PerPeerLimit} ->
+            {reply, false, State};
+        _ ->
+            riak_kv_replrtq_snk:set_worker_counts(WorkerCount, PerPeerLimit),
+            lists:foreach(
+                fun({QN, PI}) -> prompt_discovery(QN, PI, count_change) end,
+                State#state.discovery_peers),
+            {reply, true, State}
+    end.
+
+handle_cast({prompt_discovery, {QueueName, PeerInfo}}, State) ->
+    _ = prompt_discovery(QueueName, PeerInfo, regular),
+    {noreply, State}.
+
+handle_info({prompt_discovery, QueueName}, State) ->
+    {QueueName, PeerInfo} =
+        lists:keyfind(QueueName, 1, State#state.discovery_peers),
+    ok = prompt_discovery({QueueName, PeerInfo}),
+    MinDelay = 
+        app_helper:get_env(riak_kv,
+            replrtq_prompt_min_seconds,
+            ?AUTO_DISCOVERY_MINIMUM_SECONDS),
+    MaxDelay =
+        app_helper:get_env(riak_kv,
+            replrtq_prompt_max_seconds,
+            ?AUTO_DISCOVERY_MAXIMUM_SECONDS),
+    Delay = rand:uniform(max(1, MaxDelay - MinDelay)) + MinDelay,
+    erlang:send_after(Delay * 1000, self(), {prompt_discovery, QueueName}),
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%%============================================================================
+%%% Internal functions
+%%%============================================================================
+
+-spec prompt_discovery(riak_kv_replrtq_snk:queue_name(),
+            list(riak_kv_replrtq_snk:peer_info()),
+            update|regular|count_change) -> boolean().
+prompt_discovery(QueueName, PeerInfo, Type) ->
+    {SnkWorkerCount, PerPeerLimit} = riak_kv_replrtq_snk:get_worker_counts(),
+    StartDelayMS = riak_kv_replrtq_snk:starting_delay(),
+    CurrentPeers = 
+        case Type of
+            count_change ->
+                %% Ignore current peers, to update worker counts
+                [];
+            _ ->
+                lists:usort(
+                    lists:map(
+                        fun({ID, _D, H, P, Prot}) ->
+                            {ID, StartDelayMS, H, P, Prot}
+                        end,
+                        riak_kv_replrtq_snk:current_peers(QueueName)))
+        end,
+    case discover_peers(PeerInfo, StartDelayMS) of
+        CurrentPeers ->
```

Comment:
I guess this is by design, but if you run with Type `count_change` then CurrentPeers is the empty list. So in case discover_peers returns an empty list, you don't know in which of the two cases you are.

Both cases return false, but the side effects are different.

Reply:
Tried to avoid this confusion by not relying on the empty list mis-match - rather a specific mis-match between list and atom.  Commented as well.  Don't think it is super-clean still, but improved maybe.